### ISSUE TITLE
License clarification clarification

### ISSUE THREAD

**Issue by Nutzzz** [lordmulder/stdutils#21]
@lordmulder : Rather than the zlib/libpng license that NSIS and nearly all of its plugins use, you've opted for LGPL 2.1, but you have a clarification that if used as intended (an installer using this plugin as a plugin), that would be a "work that uses the Library" rather than being "a derivative of the Library."

Can you clarify that further?  My reading is that per Section 5, that the use "falls outside the scope of this License" and that the use is therefore unrestricted, but it's also possible to read it that Section 6 still controls.

As things are, the lawyer would prefer me to rewrite my NSIS script to remove your plugin.  :-(

**Comment by lordmulder** [comment]
The _GPL_ (**not** LGPL) essentially says that, if you incorporate _any_ GPL'd code into your software - even by calling the GPL'd code from a _separate(!)_ DLL file - then _your_ software becomes a "derivative work" of the GPL'd code. And, therefore, _your_ software - as a whole - must be distributed under the GPL too!

The LGPL is more permissive. It allows proprams **not** released under the LGPL, i.e. _proprietary_ programs, to use a library that is released under the LGPL - _**provided that the proprietary program and the LGPL'd library remain two separate and exchangeable entities**_. In this case, the _non_-LGPL program is considered "a work that uses the library", it is **not** "a derivative of the Library." Consequently, because your program does **not** become a "derivative" of the LGPL'd code, it can remain under its own "proprietary" license.

Now, there is some debate about when the LGPL is applicable and when not. The general consensus seems to be: If the _proprietary_ program remains in its own separate EXE or DLL file (or Shared Object on Linux), while the LGPL'd library also remains in its own separate DLL (or Shared Object), then things are fine - as long as you are distributing the LGPL'd library exactly as provided by the author. But what if a header (.h) file of the LGPL'd library needs to be _built into_ the proprietary EXE file that whishes to call the LGPL'd DLL file ???

For clarification I have added the notice that calling my plug-in, which is released under the LGPL, from an NSIS installer - which may be proprietary or not - is definitely a legitimate "use of the library". Your installer does **not** become a "derivative work" of my plug-in, just because it calls my plug-in. This, of course, requires that you distribute my plug-in exactly as I provide it. If you make any _changes_ to my plug-in, then you **are** required to share the _modified_ sources of the plug-in!

Regards,
MuldeR

**Comment by Nutzzz** [comment]
I agree that this is coming from an overabundance of caution on their part, and there is the overlay of having to deal with some contract provisions about open source material in an agreement with a third party.

One thrust of their argument, as I understand it, and that I was specifically asking for clarification on, is that--your clarification notwithstanding--because we include your unmodified DLL in a self-extracting executable it becomes a combinatorial work and therefore falls under Section 6, i.e., that the second paragraph applies: "You must give prominent notice with each copy of the work that the Library is used in it and that the Library and its use are covered by this License.  You must supply a copy of this License.  If the work during execution displays copyright notices, you must include the copyright notice for the Library among them, as well as a reference directing the user to the copy of this License."  [e.g., one concern is that there is currently a copyright notice in NSIS' branding section, so it would be difficult to fit further copyright notices in such a small space.]  This reading then also requires that we pick one from 6(a)-(e) (which in the context of an unchanged source seems rather strange).

It seems there is another argument that some portion of your code may have been lifted from another project with a more restrictive license.

Regardless, thanks so much for your work on this plugin and your attention to this question.

**Comment by lordmulder** [comment]
> One thrust of their argument, as I understand it, and that I was specifically asking for clarification on, is that--your clarification notwithstanding--because we include your unmodified DLL in a self-extracting executable it becomes a combinatorial work and therefore falls under Section 6, i.e., that the second paragraph applies: 

I don't think so! An NSIS installer, despite being a "single" EXE file, is actually composed of to destinct parts: an _EXE header_ and the compressed data that follows. The compressed data is **not** part of the actual PE/Win32 executable. It is **not** even contained in the "resource" section of the PE/Win32 executable. It simply appended to the physical end of the PE/Win32 file - just like _any_ "SFX" archive is constructed. And, most important, the plug-in DLL is **_not_** _linked_ to the NSIS EXE header in a "statical" way at all! It's contained within the compressed data part of installer file. And it will be extracted, to the `%TMP%`, as a completely separate DLL file, before it gets called by the EXE header (installer). So this is exactly the same use-case that is commonly accepted as the "LGPL use-case". Packing the NSIS EXE header and the DLL into the same file -  the installer file - is actually **no** different than wrapping an program EXE and the library DLL into the same ZIP archive for deployment. And this is what everybody does with LGPL'd libraries all the time, right? :smirk:

Anyhow, if there is still any doubt left, **_I_** am the author of the library and _**I**_ decide which license to release my code under. So, if **_I_** add a "clarification" to the license terms that explains how the LGPL applies to the specific use-case of my library, then this is what applies! Yes, there could be problem if I include any "3rd party" code that use a license which conflicts with my "clarification". But, to the best of my knowledge that is **not** the case.

> It seems there is another argument that some portion of your code may have been lifted from another project with a more restrictive license.

Well, some functions in my plug-in have been "inspired" by other plug-in's. But I did **not** "copy & paste" their code into my project. Instead, I looked at their code to learn how certain things are done. And then I implemented it myself. So, my code is **not** a "derivative" of their code. Unless what these other plug-in's do is **_patented_** - which I doubt - writing my _own_ code from the scratch that achieves the same thing in a "similar" fashion is **not** a problem at all.

The only two "3rd party" libraries I really use are **RHash** and **BLAKE2**. _RHash_ is released under the "RHash License", which they say is compatible to BSD and MIT license - so even way more permissive than LGPL. And _BLAKE2_ is even released under "CC0" license, which is a license intended to make the work as much "public domain" as possible by the individual jurisdiction.
